x/config/config.go

```diff
@@ -171,3 +174,79 @@ func NewPacketListener(transportConfig string) (transport.PacketListener, error)
 		return nil, fmt.Errorf("config scheme '%v' is not supported", url.Scheme)
 	}
 }
+
+func SanitizeConfig(transportConfig string) (string, error) {
+	// Do nothing if the config is empty
+	if transportConfig == "" {
+		return "", nil
+	}
+	// Split the string into parts
+	parts := strings.Split(transportConfig, "|")
+
+	// Iterate through each part
+	for i, part := range parts {
+		u, err := parseConfigPart(part)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse config part: %w", err)
+		}
+		scheme := strings.ToLower(u.Scheme)
+		switch scheme {
+		case "ss", "socks5", "vless":
+			parts[i], _ = sanitizeURL(u)
+		case "split", "tls", "tlsfrag":
+			// No sanitization needed
+			parts[i] = u.String()
+		default:
+			parts[i] = scheme + "://UNKNOWN"
+		}
+	}
+
+	// Join the parts back into a string
+	return strings.Join(parts, "|"), nil
+}
+
+func sanitizeURL(u *url.URL) (string, error) {
+	const redactedPlaceholder = "REDACTED"
+	scheme := strings.ToLower(u.Scheme)
+	if u.User != nil {
+		u.User = url.User(redactedPlaceholder)
+		return u.String(), nil
+	} else {
+		newURL, err := checkBase64Encoding(u)
+		if err != nil {
+			return scheme + ":" + redactedPlaceholder, err
+		}
+		if newURL.User != nil {
+			newURL.User = url.User(redactedPlaceholder)
+			return newURL.String(), nil
+		} else {
+			// If no user info is found, return the scheme and redacted placeholder
+			return scheme + ":" + redactedPlaceholder, nil
+		}
+	}
+}
+
+// checkBase64Encoding checks if url has scheme://host#fragment with host being base64
+// encoded string. iIf so, then decodes the base64 string and parses it as a url
+func checkBase64Encoding(u *url.URL) (*url.URL, error) {
```

Comment:
As discussed, the proper way to parse Shadowsocks is to define functions to parse SIP0002 and Legacy formats, and try the first one then the other, as in https://github.com/Jigsaw-Code/outline-shadowsocksconfig/blob/6d01c4b1beae3a3bc2a1074e5e310dfa023be57d/src/shadowsocks_config.ts#L206